[css-grid] Wrong 'auto' track size for a grid item

MickyG03 opened the pull request:
#### 78184d38af9fb3351dd7aa35ea45fc640ad727d7
<pre>
[css-grid] Wrong &apos;auto&apos; track size for a grid item
with a specified &apos;width:max-content&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=216144">https://bugs.webkit.org/show_bug.cgi?id=216144</a>

Reviewed by NOBODY (OOPS!).

This patch corrects the logic determining whether a grid container is
sized under a max-content constraint. The previous implementation
assumed that inline-grid implies a max-content constraint, which is
incorrect, as atomic inlines use fit-content sizing behavior.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::isUnderMaxContentConstraint const):
(WebCore::GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem):
</pre>
----------------------------------------------------------------------
#### 40c3957c5ff2c75527c788245781795cc4813f1c
<pre>
Wrong &apos;auto&apos; track size in minmax content for a grid item with a specified &apos;width:max-content&apos;
<a href="https://bugs.webkit.org/show_bug.cgi?id=216144">https://bugs.webkit.org/show_bug.cgi?id=216144</a>

Reviewed by NOBODY (OOPS!).

The addition of the new condition in sizeTrackToFitNonSpanningItem allows the grid track to grow to the max-content contribution of its items if the track has an auto min track sizing function and the grid container is being sized under a min/max-content constraint. This behavior applies only when the min track sizing function is auto.

* LayoutTests/fast/css-grid-layout/max-content-track-sizing-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/max-content-track-sizing.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::isUnderMaxContentConstraint const):
(WebCore::GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78184d38af9fb3351dd7aa35ea45fc640ad727d7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101559 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/21227 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11536 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/106717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/52192 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103599 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21535 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29722 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77347 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34371 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104566 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16627 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91716 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57685 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16451 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9733 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51539 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86321 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9810 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/109068 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28692 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/21114 "Found 2 new test failures: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-hidden-page.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/86320 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/29053 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87917 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/85882 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30622 "Passed tests") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8335 "Found 1 new test failure: imported/w3c/web-platform-tests/css/css-sizing/range-percent-intrinsic-size-1.html (failure)") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/22824 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28620 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33909 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28431 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31754 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29990 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MickyG03 committed:
Wrong 'auto' track size in minmax content for a grid item with a specified 'width:max-content'
https://bugs.webkit.org/show_bug.cgi?id=216144

Reviewed by NOBODY (OOPS!).

The addition of the new condition in sizeTrackToFitNonSpanningItem allows the grid track to grow to the max-content contribution of its items if the track has an auto min track sizing function and the grid container is being sized under a min/max-content constraint. This behavior applies only when the min track sizing function is auto.

* LayoutTests/fast/css-grid-layout/max-content-track-sizing-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/max-content-track-sizing.html: Added.
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::isUnderMaxContentConstraint const):
(WebCore::GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem):
* Source/WebCore/rendering/GridTrackSizingAlgorithm.h:

Ahmad-S792 commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -322,6 +330,14 @@ void GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem(const GridSpan& spa
+        // The min track sizing function is intrinsic (e.g., auto, min-content, max-content).
+        // If the max track sizing function is fixed (e.g., 40px), only allow growth if the min track sizing function is 'auto'.
+        if (!trackSize.hasMinContentMaxTrackBreadth() && (!trackSize.hasFixedMaxTrackBreadth() || trackSize.hasAutoMinTrackBreadth()))
+
```
Extra space.

Loirooriol commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -280,6 +280,14 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
+{
+    // inline-grid sizes to its content by default, which is a max-content constraint
+    if (m_renderGrid->style().display() == DisplayType::InlineGrid)
+        return true;
```
That's not true? The automatic size of an atomic inline is `fit-content`, which indeed is like `max-content` if the available space is big enough, but can also be `min-content` or something in between.

Also, the grid container can have a non-auto size, which would make the above moot.

MickyG03 replied:
Thanks for the feedback! You're right — my initial check for inline-grid assuming a max-content constraint was not spec-accurate.

I've updated the logic.

Also initial code was not solving the layout issue mentioned by Greg Whitworth. (https://codepen.io/gregwhitworth/pen/BarERQv) which it does now.

This should better reflect the spec behavior for both max-content and no-constraint scenarios. Let me know if you see any edge cases I might have missed.

Sincerely,
Mitesh

MickyG03 committed:
[css-grid] Wrong 'auto' track size for a grid item
with a specified 'width:max-content'
https://bugs.webkit.org/show_bug.cgi?id=216144

Reviewed by NOBODY (OOPS!).

This patch corrects the logic determining whether a grid container is
sized under a max-content constraint. The previous implementation
assumed that inline-grid implies a max-content constraint, which is
incorrect, as atomic inlines use fit-content sizing behavior.

* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::isUnderMaxContentConstraint const):
(WebCore::GridTrackSizingAlgorithm::sizeTrackToFitNonSpanningItem):

Loirooriol commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -280,6 +280,24 @@ LayoutUnit GridTrackSizingAlgorithm::initialGrowthLimit(const GridTrackSize& tra
+    if (m_direction == GridTrackSizingDirection::ForColumns) {
+        const auto& gridStyle = m_renderGrid->style();
+        // Check if width is explicitly set to max-content
+        if (gridStyle.width().isMaxContent())
```
This is still wrong, it's not a matter of the computed style, it's a matter of whether the grid is being sized under a min/max-constraint, or the size of the grid container is known and the grid is being laid out into there.

Blink tracks this with

```cpp
enum class SizingConstraint { kLayout, kMaxContent, kMinContent };
```
